### ISSUE TITLE
Failure cases for `Scan` with Numba backend when `outputs_info` mixes `None` and `TensorVariable`s

### ISSUE THREAD

**Issue by jessegrabowski** [aesara-devs/aesara#1233]
I have found two modes of failure in this case. I apologize if it's bad form to mix them in a single issue, but they seem quite related so I'm presenting them together. Here's an MRE with some test cases:

```python
x = at.dscalar('x')

def threes():
    return at.constant(3.0)

def count(x):
    return x + 1

def mixed_returns_1(x):
    return x + 1, at.constant(3.0)

def mixed_returns_2(x):
    return x + 1, x

count_result, count_updates = aesara.scan(count,
                                          outputs_info=[x],
                                          n_steps=10)

three_result, three_updates = aesara.scan(threes,
                                          outputs_info=[None],
                                          n_steps=10)

mixed_result1, mixed_updates1 = aesara.scan(mixed_returns_1,
                                          outputs_info=[x, None],
                                          n_steps=10)

mixed_result2, mixed_updates2 = aesara.scan(mixed_returns_2,
                                          outputs_info=[x, None],
                                          n_steps=10)


f_count = aesara.function([x], count_result, mode='NUMBA', updates=count_updates)
f_three = aesara.function([], three_result, mode='NUMBA', updates=three_updates)
f_mixed1 = aesara.function([x], mixed_result1, mode='NUMBA', updates=mixed_updates1)
f_mixed2 = aesara.function([x], mixed_result2, mode='NUMBA', updates=mixed_updates2)
```

The first two cases work as expected:
```
f_count(3)
>>> array([ 3.,  4.,  5.,  6.,  7.,  8.,  9., 10., 11., 12.])

f_three()
>>> array([3., 3., 3., 3., 3., 3., 3., 3., 3., 3.], dtype=float32)
```

The third case fails because (I think?) a variable is not automatically created for `at.constant(3)`.

```
f_mixed1(3)
```
<details><summary>NameError: name 'auto_15128' is not defined</summary>

<p>

```
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:202, in streamline.<locals>.streamline_default_f()
    199 for thunk, node, old_storage in zip(
    200     thunks, order, post_thunk_old_storage
    201 ):
--> 202     thunk()
    203     for old_s in old_storage:

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\basic.py:668, in JITLinker.create_jitable_thunk.<locals>.thunk(fgraph, fgraph_jit, thunk_inputs, thunk_outputs)
    662 def thunk(
    663     fgraph=self.fgraph,
    664     fgraph_jit=fgraph_jit,
    665     thunk_inputs=thunk_inputs,
    666     thunk_outputs=thunk_outputs,
    667 ):
--> 668     outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
    670     for o_var, o_storage, o_val in zip(fgraph.outputs, thunk_outputs, outputs):

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
NameError: name 'auto_15128' is not defined
During: resolving callee type: type(CPUDispatcher(<function numba_funcified_fgraph at 0x0000029591A25820>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4096101l (8)

During: resolving callee type: type(CPUDispatcher(<function numba_funcified_fgraph at 0x0000029591A25820>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4096101l (8)


File "..\..\AppData\Local\Temp\tmp4096101l", line 8:
def scan(n_steps, auto_17049_0, auto_15051_1):
    <source elided>
        inner_args = (auto_17049_0[i], )
        (auto_17049_0[i+1], auto_15051_1[i]) = numba_at_inner_func(*inner_args)
        ^

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591A25670>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmpmiu4neto (12)

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591A25670>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmpmiu4neto (12)


File "..\..\AppData\Local\Temp\tmpmiu4neto", line 12:
def numba_funcified_fgraph(x):
    <source elided>
    # forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
    auto_17051, auto_17052 = scan(auto_15051, auto_17049, auto_15051)
    ^


During handling of the above exception, another exception occurred:

TypingError                               Traceback (most recent call last)
Cell In [35], line 1
----> 1 f_mixed1(3)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\compile\function\types.py:971, in Function.__call__(self, *args, **kwargs)
    968 t0_fn = time.time()
    969 try:
    970     outputs = (
--> 971         self.vm()
    972         if output_subset is None
    973         else self.vm(output_subset=output_subset)
    974     )
    975 except Exception:
    976     restore_defaults()

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:206, in streamline.<locals>.streamline_default_f()
    204             old_s[0] = None
    205 except Exception:
--> 206     raise_with_op(fgraph, node, thunk)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:534, in raise_with_op(fgraph, node, thunk, exc_info, storage_map)
    529     warnings.warn(
    530         f"{exc_type} error does not allow us to add an extra error message"
    531     )
    532     # Some exception need extra parameter in inputs. So forget the
    533     # extra long error message in that case.
--> 534 raise exc_value.with_traceback(exc_trace)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:202, in streamline.<locals>.streamline_default_f()
    198 try:
    199     for thunk, node, old_storage in zip(
    200         thunks, order, post_thunk_old_storage
    201     ):
--> 202         thunk()
    203         for old_s in old_storage:
    204             old_s[0] = None

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\basic.py:668, in JITLinker.create_jitable_thunk.<locals>.thunk(fgraph, fgraph_jit, thunk_inputs, thunk_outputs)
    662 def thunk(
    663     fgraph=self.fgraph,
    664     fgraph_jit=fgraph_jit,
    665     thunk_inputs=thunk_inputs,
    666     thunk_outputs=thunk_outputs,
    667 ):
--> 668     outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
    670     for o_var, o_storage, o_val in zip(fgraph.outputs, thunk_outputs, outputs):
    671         compute_map[o_var][0] = True

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
NameError: name 'auto_15128' is not defined
During: resolving callee type: type(CPUDispatcher(<function numba_funcified_fgraph at 0x0000029591A25820>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4096101l (8)

During: resolving callee type: type(CPUDispatcher(<function numba_funcified_fgraph at 0x0000029591A25820>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4096101l (8)


File "..\..\AppData\Local\Temp\tmp4096101l", line 8:
def scan(n_steps, auto_17049_0, auto_15051_1):
    <source elided>
        inner_args = (auto_17049_0[i], )
        (auto_17049_0[i+1], auto_15051_1[i]) = numba_at_inner_func(*inner_args)
        ^

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591A25670>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmpmiu4neto (12)

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591A25670>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmpmiu4neto (12)


File "..\..\AppData\Local\Temp\tmpmiu4neto", line 12:
def numba_funcified_fgraph(x):
    <source elided>
    # forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
    auto_17051, auto_17052 = scan(auto_15051, auto_17049, auto_15051)
    ^

Apply node that caused the error: forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
Toposort index: 4
Inputs types: [TensorType(int8, ()), TensorType(float64, (None,)), TensorType(int8, ())]
Inputs shapes: [()]
Inputs strides: [()]
Inputs values: [array(3.)]
Outputs clients: [['output'], ['output']]

HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.

```

</p>

</details>

The final case fails due to a failed item assignment. It seems that the return variable associated with the `None` is an array inside a list? 

```
f_mixed2(3)
```

<details><summary>No implementation of function Function(<built-in function setitem>) found for signature: >>> setitem(list(array(float64, 0d, C))<iv=None>, int32, float64)</summary>

<p>

```
---------------------------------------------------------------------------
TypingError                               Traceback (most recent call last)
File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:202, in streamline.<locals>.streamline_default_f()
    199 for thunk, node, old_storage in zip(
    200     thunks, order, post_thunk_old_storage
    201 ):
--> 202     thunk()
    203     for old_s in old_storage:

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\basic.py:668, in JITLinker.create_jitable_thunk.<locals>.thunk(fgraph, fgraph_jit, thunk_inputs, thunk_outputs)
    662 def thunk(
    663     fgraph=self.fgraph,
    664     fgraph_jit=fgraph_jit,
    665     thunk_inputs=thunk_inputs,
    666     thunk_outputs=thunk_outputs,
    667 ):
--> 668     outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
    670     for o_var, o_storage, o_val in zip(fgraph.outputs, thunk_outputs, outputs):

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function setitem>) found for signature:
 
 >>> setitem(list(array(float64, 0d, C))<iv=None>, int32, float64)
 
There are 16 candidate implementations:
    - Of which 14 did not match due to:
    Overload of function 'setitem': File: <numerous>: Line N/A.
      With argument(s): '(list(array(float64, 0d, C))<iv=None>, int32, float64)':
     No match.
    - Of which 2 did not match due to:
    Overload in function 'SetItemSequence.generic': File: numba\core\typing\collections.py: Line 56.
      With argument(s): '(list(array(float64, 0d, C))<iv=None>, int32, float64)':
     Rejected as the implementation raised a specific error:
       TypingError: invalid setitem with value of float64 to element of array(float64, 0d, C)
  raised from C:\Users\Jesse\miniconda3\envs\econ\lib\site-packages\numba\core\typing\collections.py:65

During: typing of setitem at C:\Users\Jesse\AppData\Local\Temp\tmpjararahe (8)

File "..\..\AppData\Local\Temp\tmpjararahe", line 8:
def scan(n_steps, auto_17955_0, auto_15146_1):
    <source elided>
        inner_args = (auto_17955_0[i], )
        (auto_17955_0[i+1], auto_15146_1[i]) = numba_at_inner_func(*inner_args)
        ^

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591450F70>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4zfz_tx0 (12)

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591450F70>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4zfz_tx0 (12)


File "..\..\AppData\Local\Temp\tmp4zfz_tx0", line 12:
def numba_funcified_fgraph(x):
    <source elided>
    # forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
    auto_17957, auto_17958 = scan(auto_15146, auto_17955, auto_15146)
    ^


During handling of the above exception, another exception occurred:

TypingError                               Traceback (most recent call last)
Cell In [33], line 1
----> 1 f_mixed2(3)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\compile\function\types.py:971, in Function.__call__(self, *args, **kwargs)
    968 t0_fn = time.time()
    969 try:
    970     outputs = (
--> 971         self.vm()
    972         if output_subset is None
    973         else self.vm(output_subset=output_subset)
    974     )
    975 except Exception:
    976     restore_defaults()

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:206, in streamline.<locals>.streamline_default_f()
    204             old_s[0] = None
    205 except Exception:
--> 206     raise_with_op(fgraph, node, thunk)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:534, in raise_with_op(fgraph, node, thunk, exc_info, storage_map)
    529     warnings.warn(
    530         f"{exc_type} error does not allow us to add an extra error message"
    531     )
    532     # Some exception need extra parameter in inputs. So forget the
    533     # extra long error message in that case.
--> 534 raise exc_value.with_traceback(exc_trace)

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\utils.py:202, in streamline.<locals>.streamline_default_f()
    198 try:
    199     for thunk, node, old_storage in zip(
    200         thunks, order, post_thunk_old_storage
    201     ):
--> 202         thunk()
    203         for old_s in old_storage:
    204             old_s[0] = None

File ~\miniconda3\envs\econ\lib\site-packages\aesara\link\basic.py:668, in JITLinker.create_jitable_thunk.<locals>.thunk(fgraph, fgraph_jit, thunk_inputs, thunk_outputs)
    662 def thunk(
    663     fgraph=self.fgraph,
    664     fgraph_jit=fgraph_jit,
    665     thunk_inputs=thunk_inputs,
    666     thunk_outputs=thunk_outputs,
    667 ):
--> 668     outputs = fgraph_jit(*[x[0] for x in thunk_inputs])
    670     for o_var, o_storage, o_val in zip(fgraph.outputs, thunk_outputs, outputs):
    671         compute_map[o_var][0] = True

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:468, in _DispatcherBase._compile_for_args(self, *args, **kws)
    464         msg = (f"{str(e).rstrip()} \n\nThis error may have been caused "
    465                f"by the following argument(s):\n{args_str}\n")
    466         e.patch_message(msg)
--> 468     error_rewrite(e, 'typing')
    469 except errors.UnsupportedError as e:
    470     # Something unsupported is present in the user code, add help info
    471     error_rewrite(e, 'unsupported_error')

File ~\miniconda3\envs\econ\lib\site-packages\numba\core\dispatcher.py:409, in _DispatcherBase._compile_for_args.<locals>.error_rewrite(e, issue_type)
    407     raise e
    408 else:
--> 409     raise e.with_traceback(None)

TypingError: Failed in nopython mode pipeline (step: nopython frontend)
Failed in nopython mode pipeline (step: nopython frontend)
No implementation of function Function(<built-in function setitem>) found for signature:
 
 >>> setitem(list(array(float64, 0d, C))<iv=None>, int32, float64)
 
There are 16 candidate implementations:
    - Of which 14 did not match due to:
    Overload of function 'setitem': File: <numerous>: Line N/A.
      With argument(s): '(list(array(float64, 0d, C))<iv=None>, int32, float64)':
     No match.
    - Of which 2 did not match due to:
    Overload in function 'SetItemSequence.generic': File: numba\core\typing\collections.py: Line 56.
      With argument(s): '(list(array(float64, 0d, C))<iv=None>, int32, float64)':
     Rejected as the implementation raised a specific error:
       TypingError: invalid setitem with value of float64 to element of array(float64, 0d, C)
  raised from C:\Users\Jesse\miniconda3\envs\econ\lib\site-packages\numba\core\typing\collections.py:65

During: typing of setitem at C:\Users\Jesse\AppData\Local\Temp\tmpjararahe (8)

File "..\..\AppData\Local\Temp\tmpjararahe", line 8:
def scan(n_steps, auto_17955_0, auto_15146_1):
    <source elided>
        inner_args = (auto_17955_0[i], )
        (auto_17955_0[i+1], auto_15146_1[i]) = numba_at_inner_func(*inner_args)
        ^

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591450F70>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4zfz_tx0 (12)

During: resolving callee type: type(CPUDispatcher(<function scan at 0x0000029591450F70>))
During: typing of call at C:\Users\Jesse\AppData\Local\Temp\tmp4zfz_tx0 (12)


File "..\..\AppData\Local\Temp\tmp4zfz_tx0", line 12:
def numba_funcified_fgraph(x):
    <source elided>
    # forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
    auto_17957, auto_17958 = scan(auto_15146, auto_17955, auto_15146)
    ^

Apply node that caused the error: forall_inplace,cpu,scan_fn}(TensorConstant{10}, IncSubtensor{InplaceSet;:int64:}.0, TensorConstant{10})
Toposort index: 4
Inputs types: [TensorType(int8, ()), TensorType(float64, (None,)), TensorType(int8, ())]
Inputs shapes: [()]
Inputs strides: [()]
Inputs values: [array(3.)]
Outputs clients: [['output'], ['output']]

HINT: Re-running with most Aesara optimizations disabled could provide a back-trace showing when this node was created. This can be done by setting the Aesara flag 'optimizer=fast_compile'. If that does not work, Aesara optimizations can be disabled with 'optimizer=None'.
HINT: Use the Aesara flag `exception_verbosity=high` for a debug print-out and storage map footprint of this Apply node.

```

</p>

</details>

I have this list-wrapping hypothesis because I can re-create nearly the exact error message with the following code snippet:

```python
@njit
def set_item(l, i, x):
    l[i] = x
    return l

set_item([np.zeros(())], 0, 3)
```

Note that `aesara.gradient.jacobian` fails out-of-the-box with the final case (invalid indexing) when you try to compile it to numba.

**Comment by brandonwillard** [comment]
The `NameError` issue appears to be a more general oversight in `fgraph_to_python`: it doesn't process `Constant`-only outputs.

**Comment by brandonwillard** [comment]
The second issue is fixed in #1203.
